File: src/interval.ts
export function interval(min: number, max: number) {
  return new Interval(min, max);
}

export class Interval {
  constructor(
    public min: number,
    public max: number,
  ) {}

  get size() {
    return this.max - this.min;
  }

  contains(x: number) {
    return this.min <= x && x <= this.max;
  }

  surrounds(x: number) {
    return this.min < x && x < this.max;
  }

  clamp(x: number) {
    if (x < this.min) return this.min;
    if (x < this.max) return this.max;
    return x;
  }

  static empty = new Interval(Infinity, -Infinity);
  static universe = new Interval(-Infinity, +Infinity);
}

File: raytracer/interval.go
package raytracer

import "math"

type Interval struct {
	Min float64
	Max float64
}

var (
	EmptyInterval    = Interval{Min: math.MaxFloat64, Max: -math.MaxFloat64}
	UniverseInterval = Interval{Min: -math.MaxFloat64, Max: math.MaxFloat64}
)

func (iv Interval) Size() float64 {
	return iv.Max - iv.Min
}

func (iv Interval) Contains(v float64) bool {
	return iv.Min <= v && v <= iv.Max
}

func (iv Interval) Surrounds(v float64) bool {
	return iv.Min < v && v < iv.Max
}

func (iv Interval) Clamp(v float64) float64 {
	if v < iv.Min {
		return iv.Min
	}
	if v > iv.Max {
		return iv.Max
	}
	return v
}
